src/luarocks/deps.tl

```diff
@@ -324,28 +326,28 @@ function deps.fulfill_dependencies(rockspec, depskey, deps_mode, verify, deplock
       return nil, err
    end
 
-   deps.report_missing_dependencies(name, version, rockspec[depskey], deps_mode, rocks_provided)
+   deps.report_missing_dependencies(name, version, (rockspec as {string: Dependencies})[depskey].queries, deps_mode, rocks_provided)
 
    util.printout()
 
    local get_versions = prepare_get_versions(deps_mode, rocks_provided, depskey)
-   for _, dep in ipairs(rockspec[depskey]) do
+   for _, depq in ipairs((rockspec as {string: Dependencies})[depskey].queries) do
 
       util.printout(("%s %s depends on %s (%s)"):format(
-         name, version, tostring(dep), rock_status(dep, get_versions)))
+         name, version, tostring(depq), rock_status(depq, get_versions)))
 
-      local ok, found_or_err, _, no_upgrade = deps.fulfill_dependency(dep, deps_mode, rocks_provided, verify, depskey)
-      if ok then
-         deplocks.add(depskey, dep.name, found_or_err)
+      local okfulfill, found_or_err, _ = deps.fulfill_dependency(depq, deps_mode, rocks_provided, verify, depskey)
+      if okfulfill then
+         deplocks.add(depskey, depq.name, found_or_err)
       else
-         if no_upgrade then
-            util.printerr("This version of "..name.." is designed for use with")
-            util.printerr(tostring(dep)..", but is configured to avoid upgrading it")
-            util.printerr("automatically. Please upgrade "..dep.name.." with")
-            util.printerr("   luarocks install "..dep.name)
-            util.printerr("or look for a suitable version of "..name.." with")
-            util.printerr("   luarocks search "..name)
-         end
+         -- if no_upgrade then
+         --    util.printerr("This version of "..name.." is designed for use with")
+         --    util.printerr(tostring(dep)..", but is configured to avoid upgrading it")
+         --    util.printerr("automatically. Please upgrade "..dep.name.." with")
+         --    util.printerr("   luarocks install "..dep.name)
+         --    util.printerr("or look for a suitable version of "..name.." with")
+         --    util.printerr("   luarocks search "..name)
+         -- end
```

Comment:
oh wow, looks like this codepath was broken this whole time! I think I have a fix.

Reply:
I pushed a fix, just leaving this open here so you can see it